### PR TITLE
Refactoring cli.py and LinterError

### PR DESCRIPTION
**Pull request by ghost**
Refactoring cli.py, LinterError and logging.

### CODE REVIEW

File: j2lint/linter/runner.py [hunk 1]
```diff
@@ -68,6 +68,4 @@ def run(self):
         # Update list of checked files
         self.checked_files.update([file_dict['path'] for file_dict in files])
 
-        # TODO - log also warnings
-        logger.info("Linting errors found %s", errors)
```
Comment (gmuloc) [inline j2lint/linter/runner.py:72]: ~~Hmmm the comment was to also log the warnings and not only the errors~~

Ok ignore this, you moved it to collection

File: j2lint/linter/error.py [hunk 2]
```diff
@@ -18,24 +16,31 @@ def __init__(self, line_number, line, filename, rule, message=None):
...
+        """ setting json output format"""
+        logger.debug("JSON output enabled")
+        error = json.dumps({"id": self.rule.id,
+                            "message": self.message,
+                            "filename": self.filename,
+                            "line_number": self.line_number,
+                            "line": self.line,
+                            "severity": self.rule.severity
+                            })
         return error
```
Comment (gmuloc) [inline j2lint/linter/error.py]:
```suggestion
    def to_json(self):
        """ setting json output format"""
        logger.debug("JSON output enabled")
        return = json.dumps({"id": self.rule.id,
                            "message": self.message,
                            "filename": self.filename,
                            "line_number": self.line_number,
                            "line": self.line,
                            "severity": self.rule.severity
                            })
```

File: j2lint/linter/error.py [hunk 3]
```diff
@@ -18,24 +16,31 @@ def __init__(self, line_number, line, filename, rule, message=None):
...
-            error = format_str.format(self.rule.id, self.rule.description,
-                                      self.filename, self.line_number, self.line,
-                                      self.message, self.rule.short_description)
+            logger.debug("Verbose mode enabled")
+            format_str = ("Linting rule: {0}\nRule description: "
+                          "{1}\nError line: {2}:{3} {4}\nError message: {5}\n")
+        error = format_str.format(self.rule.id, self.rule.description,
+                                  self.filename, self.line_number, self.line,
+                                  self.message, self.rule.short_description)
+        return error
```
Comment (gmuloc) [inline j2lint/linter/error.py]:
```suggestion
        return format_str.format(self.rule.id, self.rule.description,
                                 self.filename, self.line_number, self.line,
                                 self.message, self.rule.short_description)
```

File: j2lint/cli.py [hunk 4]
```diff
@@ -112,6 +111,19 @@ def get_linting_issues(file_or_dir_names, options, collection, checked_files):
...
+
+
+def print_issues(key, sorted_issues, verbose):
+    """ print non-json output """
+    print(f"************ File {key}")
+    for j2_issue in sorted_issues:
+        print(f"{j2_issue.to_string(verbose)}")
+
+
 def sort_and_print_issues(options, lint_issues, issue_type, json_output):
```
Comment (gmuloc) [inline j2lint/cli.py]: We should not have a function called "sort_AND_print". 

So basically we should have:

1. get_linting_issues returning a tuple (lint_errors, lint_warnings), already sorted
2. then in the run function:
```
if json:
  print_json(lint_errors, lint_warnings)
else:
  print_string(lint_errors, lint_warnings)
```

Comment (Shivani-chourasiya) [reply j2lint/cli.py]: I think we should keep this function, as we are using it for errors and warnings both. sorting part can be in get_linting_issues() but for printing and getting total_errors and total_warnings we need to make separate loop for both errors and warnings.
for errors- 
   ```
 for key, errors in lint_errors.items():
        if not errors:
            continue
        else:
            print(f"************ File {key}")
            for error in errors:
                print(error.to_string(verbose))
```
  for warnings- 
   ```
 for key, warnings in lint_warnings.items():
        if not warnings:
            continue
        else:
            print(f"************ File {key}")
            for warning in warnings:
                print(warning.to_string(verbose))
```

Comment (gmuloc) [reply j2lint/cli.py]: a function should have only one role - being call `sort_and_print` is an anti pattern. We are already doing both loops today so it does not change


Here are some functions respecting the separation of the responsabilities (need to add docstring and verify the code runs):

```python
def sort_issues(issues):
    """Sorted list of issues

    Args:
        issues (list): list of issue dictionaries

    Returns:
        list: list of sorted issue dictionaries
    """
    issues.sort(
        key=lambda issue: (
            issue.filename,
            issue.line_number,
            issue.rule.id
        )
    )
    return issues


def get_linting_issues(file_or_dir_names, options, collection, checked_files):
    """ checking errors and warnings """
    lint_errors = {}
    lint_warnings = {}
    files = get_files(file_or_dir_names)

    # Get linting issues
    for file_name in files:
        runner = Runner(collection, file_name, checked_files)
        if file_name not in lint_errors:
            lint_errors[file_name] = []
        if file_name not in lint_warnings:
            lint_warnings[file_name] = []
        j2_errors, j2_warnings = runner.run()
        lint_errors[file_name].extend(j2_errors)
        lint_warnings[file_name].extend(j2_warnings)

    return sort_issues(lint_errors), sort_issues(lint_warnings)

def print_json(lint_errors, lint_warnings):
    """
    build and print json
    issues are already sorted
    retur total_errors, total_warnings
    """
    json_output = {}
    for key, issues in lint_errors.items():
        json_output["ERRORS"] = ([json.loads(issue.to_json()) for issue in issues])
    for key, issues in lint_warnings.items():
        json_output["WARNINGS"] = ([json.loads(issue.to_json()) for issue in issues])

    print(json.dumps(json_output))
    return len(json_output["ERRORS"]), len(json_output["WARNINGS"])

def print_string(lint_errors, lint_warnings, verbose=False):
    """
    return count
    """
    total_lint_errors = sum([len(issues) for _, issues in lint_errors.items()])
    total_lint_warnings = sum([len(issues) for _, issues in lint_warnings.items()])

    if not(total_lint_errors + total_lint_warnings):
        # no error nor warning
        print("Linting complete. No problems found.")
        return total_lint_errors, total_warnings

    print(f"Jinja2 linting finished with "
          f"{total_lint_errors} issue(s) and {total_lint_warnings} warning(s)")

    if total_lint_errors:
        print(f"\nJINJA2 LINT ERRORS")
        for key, issues in lint_errors.items():
            print(f"************ File {key}")
            for j2_issue in sorted_issues:
                print(j2_issue.to_string(verbose))
    if total_lint_warnings:
        print(f"\nJINJA2 LINT WARNINGS")
        for key, issues in lint_warnings.items():
            print(f"************ File {key}")
            for j2_issue in sorted_issues:
                print(j2_issue.to_string(verbose))

    return total_lint_errors, total_lint_warnings
```


which can be called in `run` as:

```python
def run...
    [...]
    lint_errors, lint_warnings = get_linting_issues(file_or_dir_names, options, collection, checked_files))
    if options.json:
        total_errors, total_warnings = print_json(lint_errors, lint_warnings)
    else:
        total_errors, total_warnings = print_string(lint_erros, lint_warnings, options.verbose)

    # Remove temporary file
    remove_temporary_file(stdin_filename)

    if total_errors:
        return 2

    return 0
```

**Review by gmuloc — APPROVED** [review]
LGTM - thanks for the good work!

**Pull request merged**